plugins/deisa/deisa.cxx

```diff
@@ -0,0 +1,228 @@
+#include <vector>
+
+#include <pybind11/embed.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <mpi.h>
+#include <pdi.h>
+#include <pdi/array_datatype.h>
+#include <pdi/context.h>
+#include <pdi/data_descriptor.h>
+#include <pdi/datatype.h>
+#include <pdi/expression.h>
+#include <pdi/logger.h>
+#include <pdi/paraconf_wrapper.h>
+#include <pdi/plugin.h>
+#include <pdi/python/tools.h>
+#include <pdi/ref_any.h>
+#include <pdi/scalar_datatype.h>
+
+namespace {
+
+namespace py = pybind11;
+
+using namespace PDI;
+using namespace py::literals;
+using pydict = py::dict;
+using pymod = py::module;
+using pyobj = py::object;
+
+/** The deisa plugin
+ */
+class deisa_plugin: public Plugin
+{
+	static constexpr char PYTHON_LIBRARY_COMPATIBILITY[] = "0.3.2"; // Used to check compatibility with Deisa's python library
+
+	bool m_interpreter_initialized_in_plugin = false; // Determine if python interpreter is initialized by the plugin
+	Expression m_scheduler_info;
+	unordered_map<std::string, Datatype_template_sptr> m_deisa_arrays;
+	Expression m_rank;
+	Expression m_size;
+	Expression m_time_step;
+
+public:
+	static std::pair<std::unordered_set<std::string>, std::unordered_set<std::string>> dependencies() { return {{"mpi"}, {"mpi"}}; }
+
+	deisa_plugin(Context& ctx, PC_tree_t conf)
+		: Plugin{ctx}
+	{
+		if (!Py_IsInitialized()) {
+			py::initialize_interpreter();
+			py::exec("bridge = None"); // needed because check in dtor.
+			m_interpreter_initialized_in_plugin = true;
+		}
+		check_compatibility();
+		// init params
+		each(conf, [&](PC_tree_t key_tree, PC_tree_t value) {
+			std::string key = to_string(key_tree);
+			if (key == "scheduler_info") {
+				m_scheduler_info = to_string(value);
+			} else if (key == "deisa_arrays") {
+				each(value, [&](PC_tree_t key_map, PC_tree_t value_map) { m_deisa_arrays.emplace(to_string(key_map), ctx.datatype(value_map)); });
+			} else if (key == "time_step") {
+				m_time_step = to_string(value);
+			} else if (key == "map_in") {
+				//
+			} else if (key == "logging" || key == "init_on") {
+				//
+			} else {
+				throw Config_error{key_tree, "Unknown key in Deisa file configuration: '{}'", key};
+			}
+		});
+
+
+		int mpi_size;
+		m_rank = Expression{Ref_r{ctx.desc("MPI_COMM_WORLD_rank").ref()}.scalar_value<long>()};
+		MPI_Comm comm = *static_cast<const MPI_Comm*>(Ref_r{ctx.desc("MPI_COMM_WORLD").ref()}.get());
+		MPI_Comm_size(comm, &mpi_size);
+		m_size = Expression{static_cast<long>(mpi_size)};
+
+
+		// plugin init
+		PC_tree_t init_tree = PC_get(conf, ".init_on");
+		if (!PC_status(init_tree)) {
+			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(); }, to_string(init_tree));
+		} else {
+			throw Config_error{conf, "Deisa plugin requires init_on key "};
+		}
+
+		// map_in
+		PC_tree_t map_tree = PC_get(conf, ".map_in");
+		if (!PC_status(map_tree)) {
+			each(map_tree, [&](PC_tree_t key_map, PC_tree_t value_map) {
```

Comment:
Once https://github.com/pdidev/pdi/issues/481 is fixed, we should use the `or_none` variant here.